tango/step_caches/remote_step_cache.py

```diff
@@ -0,0 +1,199 @@
+class RemoteStepCache(LocalStepCache):
+    """
+    This is a :class:`~tango.step_cache.StepCache` that's used by :class:`RemoteWorkspace`.
+    It stores the results of steps on some RemoteWorkspace.
+
+    It also keeps a limited in-memory cache as well as a local backup on disk, so fetching a
+    step's resulting subsequent times should be fast.
+
+    .. tip::
+        All remote step caches inherit from this.
+    """
+
+    Constants = RemoteConstants
+
+    def __init__(self, local_dir: Path):
+        super().__init__(local_dir)
+
+    @property
+    @abstractmethod
+    def client(self) -> RemoteClient:
+        raise NotImplementedError()
+
+    def _step_result_remote(self, step: Union[Step, StepInfo]) -> Optional[RemoteDataset]:
+        """ """
+        try:
+            dataset = self.client.get(self.Constants.step_dataset_name(step))
+            return dataset if dataset.committed else None
+        except RemoteDatasetNotFound:
+            return None
+
+    def _sync_step_remote(self, step: Step, objects_dir: Path) -> RemoteDataset:
+        dataset_name = self.Constants.step_dataset_name(step)
+        try:
+            self.client.create(dataset_name, commit=False)
+        except RemoteDatasetConflict:
+            pass
+        try:
+            self.client.sync(dataset_name, objects_dir)
+            self.client.commit(dataset_name)
+        except RemoteDatasetWriteError:
+            pass
+
+        return self.client.get(dataset_name)
+
+    def _fetch_step_remote(self, step_result, target_dir: PathOrStr) -> None:
+        try:
+            self.client.fetch(step_result, target_dir)
+        except RemoteDatasetNotFound:
+            self._raise_remote_not_found()
+
+    def __len__(self):
+        # NOTE: lock datasets should not count here. They start with the same prefix,
+        # but they never get committed.
+        return sum(
+            1
+            for ds in self.client.datasets(
+                uncommitted=False, match=self.Constants.STEP_DATASET_PREFIX
+            )
+            if ds.name is not None and ds.name.startswith(self.Constants.STEP_DATASET_PREFIX)
+        )
+
+    def _step_results_dir(self) -> str:
+        return self.Constants.STEP_RESULT_DIR
+
+    def _raise_remote_not_found(self):
+        raise RemoteNotFoundError()
+
+    def _acquire_step_lock_file(self, step: Union[Step, StepInfo], read_only_ok: bool = False):
+        return FileLock(
+            self.step_dir(step).with_suffix(".lock"), read_only_ok=read_only_ok
+        ).acquire_with_updates(desc=f"acquiring step cache lock for '{step.unique_id}'")
+
+    def __contains__(self, step: Any) -> bool:
+        if isinstance(step, (Step, StepInfo)):
+            cacheable = step.cache_results if isinstance(step, Step) else step.cacheable
+            if not cacheable:
+                return False
+
+            # TODO: old beaker step cache seems to not check locally. Should it?
```

Comment:
Yes, but out of scope for this PR.

Reply:
To be clear, they all should check locally as an optimization, but adding that to Beaker is out of scope.

Reply:
Well, actually, with these abstractions, beaker's step cache does check locally now.